gateway/mediaserver.go

```diff
@@ -0,0 +1,256 @@
+package gateway
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha1"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/smithy-go/logging"
+
+	"github.com/matterbridge-org/matterbridge/bridge/config"
+	"github.com/sirupsen/logrus"
+)
+
+type mediaServer interface {
+	handleFilesUpload(fi *config.FileInfo) (string, error)
+}
+
+type commonMediaServer struct {
+	logger *logrus.Entry
+}
+
+type httpPutMediaServer struct {
+	commonMediaServer
+
+	httpUploadPath     string
+	httpDownloadPrefix string
+}
+
+type localMediaServer struct {
+	commonMediaServer
+
+	localPath          string
+	httpDownloadPrefix string
+}
+
+type s3MediaServer struct {
+	commonMediaServer
+
+	s3client       *s3.Client
+	bucket         string
+	uploadPrefix   string
+	downloadPrefix string
+}
+
+var _ mediaServer = (*httpPutMediaServer)(nil)
+var _ mediaServer = (*localMediaServer)(nil)
+var _ mediaServer = (*s3MediaServer)(nil)
+
+func simpleS3Config(access_key, secret_access_key string, endpoint_url string) aws.Config {
+	return aws.Config{
+		Region:       "custom",
+		Credentials:  credentials.NewStaticCredentialsProvider(access_key, secret_access_key, ""),
+		Logger:       logging.Nop{},
+		BaseEndpoint: aws.String(endpoint_url),
+	}
+}
+
+func createS3MediaServer(bg *config.BridgeValues, parsed *url.URL, logger *logrus.Entry) (*s3MediaServer, error) {
+	optionsFromURL := parsed.Query()
+	secretAccessKey, _ := parsed.User.Password()
+	pathSplitted := strings.Split(strings.TrimLeft(parsed.Path, "/"), "/")
+
+	useSSL, err := strconv.ParseBool(optionsFromURL.Get("useSSL"))
+	if err != nil {
+		logger.Warn("error while parsing useSSL boolean, assuming false: ", err)
+		useSSL = false
+	}
+
+	pathStyle := true
+	if optionsFromURL.Has("pathStyle") {
+		pathStyle, err = strconv.ParseBool(optionsFromURL.Get("pathStyle"))
+		if err != nil {
+			logger.Warn("error while parsing pathStyle boolean, assuming false: ", err)
+			pathStyle = false
+		}
+	}
+
+	if len(pathSplitted) == 0 {
+		return nil, fmt.Errorf("no bucket specified")
```

Comment:
This could be solved by using static errors (maybe in combination with [wrapping](https://go.dev/blog/go1.13-errors#wrapping-errors-with-w) if we actually need the "-f" part of Errorf)